### PR TITLE
Implementation of algorithm one from the paper

### PR DESCRIPTION
**Pull request by rcmalli**
This PR is the initial effort for implementing Algorithm one for online learning using Warpgrad. I started analysing the implementation of algorithm 2. Since online learning algorithm does not require to store datapoints and model states in the buffer, I have reused [step](https://github.com/flennerhag/warpgrad/blob/d9ef72af10eec62ae92bc24595cb1a4a0207e319/src/warpgrad/warpgrad/utils.py#L58-L103) function from `warpgrad.utils` inside inner training loop.

Summary of changes:

- New wrapper for online algorithm added. This reuses functions from `warpgrad.utils`
- Simple updater class is added. However, it works only as placeholder and does nothing in the backward pass call. I am not sure if `leap` based initialization should be applied also for online learning.
- `step` function is called inside `run_batches` function of the wrapper class for each `k` times of inner update.
- Generated losses are accumulated using `meta_loss` property of wrapper class.

### CODE REVIEW

File: src/omniglot/wrapper.py [hunk 1]
```diff
@@ -139,8 +140,182 @@ def run_batches(self, batches, optimizer, train=False, meta_train=False):
...
+        """Run model on a given task, first adapting and then evaluating"""
+        self.model.no_collect()
+
+        optimizer = None
+        if train:
+            # TODO: Discuss implementation and correct it.
+            # This line breakes gradient computation for now
+            # meta_layers required_grad properties are set to False if
+            # we call init_adaptation
+            # self.model.init_adaptation()
```
Comment (rcmalli) [inline src/omniglot/wrapper.py]: Calling `self.model.init_adaptation()` produces error  when calling `backward()` at the end of each `meta_batch` since it sets meta_layer's require_grad properties to `False`. This may need us to freeze/unfreeze meta_layers in a more controlled way.

File: src/omniglot/wrapper.py [hunk 2]
```diff
@@ -139,8 +140,182 @@ def run_batches(self, batches, optimizer, train=False, meta_train=False):
...
+        ##
+        results = AggRes(results)
+
+        # Meta gradient step
+        if meta_train:
+            # at the end of collection for K steps N tasks we do the backward
+            # pass.
+            backward(self.meta_loss, self.model.meta_parameters(
+                include_init=False))
+            self._final_meta_update()
```
Comment (rcmalli) [inline src/omniglot/wrapper.py]: When we collect `k` times inner iteration for `N` tasks we can call backward pass calculate gradients.

File: src/omniglot/wrapper.py [hunk 3]
```diff
@@ -139,8 +140,182 @@ def run_batches(self, batches, optimizer, train=False, meta_train=False):
...
+                opt.zero_grad()
+                outer_input, outer_target = next(iter(batches))
+                l_outer, (l_inner, a1, a2) = step(
+                    criterion=self.criterion,
+                    x_inner=inner_input, x_outer=outer_input,
+                    y_inner=inner_target, y_outer=outer_target,
+                    model=self.model,
+                    optimizer=opt, scorer=None)
+                self.meta_loss = self.meta_loss + l_outer
+                del l_inner, a1, a2
```
Comment (rcmalli) [inline src/omniglot/wrapper.py]: These lines are calculating `outer_loss` at each state of model parameter \theta_{k}^{\tau}. However, I am not sure about how should we handle freezing and unfreezing meta and adaptation layers.

According to pseudocode, gradients of \theta_{0} must be collected using \theta_{0: k}^{\tau}. How should we implement it correctly?

File: src/warpgrad/warpgrad/updaters.py [hunk 4]
```diff
@@ -22,8 +22,46 @@
...
+        Arguments:
+            model (Warp): warped model to backprop through.
+            step_fn (function): step function for the meta gradient.
+            **opt_kwargs (kwargs): optional arguments to inner optimizer.
+        """
+
+        # init_objective = INIT_OBJECTIVES[self.init_objective]
+        # init_objective(model.named_init_parameters(suffix=None),
+        #                params, self.norm, self.bsz, step_fn)
+        pass
```
Comment (rcmalli) [inline src/warpgrad/warpgrad/updaters.py:62]: I have commented out initialization objective for now. Should we also use leap based initialization for online learning?
